redis: fix hostname resolution bug introduced in #19541

huntsman90 opened the pull request:
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message:
In #19541 , one of the latter commits changed the `hostname_resolution_required_cnt` from a simple `uint64_t` to a `shared_ptr<uint64_t>`. This meant that we didn't need to pass the variable by reference into the capture-list of the DNS resolution lambda. However, this was overlooked, and results in segfaults due to `shared_ptr` semantics.

Simply passing this by value solves the problem.

Testing:
Unsure why this didn't fail earlier, and what test scenarios would catch it. Happy to add a test as per reviewer suggestions.

mattklein123 commented:
Hmm it makes me uncomfortable this was not caught with tests. I think the reason must be that the only tests you have either only resolve a single host and/or all resolutions take place inline? Can you make sure you have a test that covers > 1 host with resolution not happening inline? I think then you should hit this, at least in ASAN. Thank you.

/wait

huntsman90 commented:
Hmm.. I did have a test earlier, but it was with one primary and two replicas. I just added a test with two primaries and two replicas each. However, I noticed that the new test I added passes even without the business logic change. The logs show that the resolution all take place serially. So, how can I force parallel resolution to reproduce the issue with old code?

mattklein123 commented:
If your test resolves localhost, it will happen inline. You might need to use mocks and do it in a unit test? This is unfortunately an issue with using localhost for testing DNS.

huntsman90 commented on test/extensions/clusters/redis/redis_cluster_test.cc:
```diff
@@ -778,6 +856,46 @@ TEST_F(RedisClusterTest, AddressAsHostname) {
   EXPECT_EQ(0U, cluster_->info()->stats().update_failure_.value());
 }
 
+TEST_F(RedisClusterTest, AddressAsHostnameMultiplePrimaries) {
```
@mattklein123 this is the unit test I've added. Do you have example of prior art on how to mock non-localhost machine in the response?

huntsman90 replied:
Thank you for the pointer! I've updated the test to reflect checks for actual asynchronous resolution. I simplified the setup a bit to reduce test complexity. The test now initializes two slots with just primaries, but both their resolutions happen asynchronously. I verified that the test fails on `main` branch, and reproduces the SEGFAULT that we'd seen in our manual testing.

mattklein123 reviewed (approved):
Awesome, thanks!
